### PR TITLE
PinotConfigProvider interface

### PR DESCRIPTION
**Pull request by npawar**
Adding a PinotConfigProvider interface, to get TableConfig and Schema, and allow listeners to register and get notified of changes. 
We already have an implementation of this in the form of TableCache in `pinot-common`. However, this impl cannot be used as is in our plugins, since plugins do not take the dependency on `pinot-common`. Accessing configs is a very common operation in plugins (e.g. minion plugins do this frequently).
Adding this interface will help us use this in our plugins and keep the dependency clean.

Adding the Listeners to the ConfigProvider, because
1. we can eliminate the need for restart/API invocation in case of certain operations like reload.
2. it's in general a good API to have on a ConfigProvider

### CODE REVIEW

**Review by mcvsubbu — COMMENTED** [review]
thanks for the PR.

Please give me a few days to think about this interface before merging, thanks.

**Review by klsince — APPROVED** [review]
neat!

**Comment by npawar** [comment]
@mayankshriv please help review..
@mcvsubbu would you be able to help review this today? I'd like to get this in today if possible

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/provider/PinotConfigProvider.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+package org.apache.pinot.spi.config.provider;
+
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.Schema;
+
+
+/**
+ * An interface for the provider of pinot configs
+ */
+public interface PinotConfigProvider {
```
Comment (mayankshriv) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/provider/PinotConfigProvider.java:28]: I take it that this interface is specific to Table config and schema, and not necessarily going to be extended to other types of configs in future? If so, perhaps we can document that in the interface.

**Review by mayankshriv — APPROVED** [review]
Thanks @npawar for creating this interface.

**Comment by mayankshriv** [comment]
> @mayankshriv please help review.. @mcvsubbu would you be able to help review this today? I'd like to get this in today if possible

@npawar Approved with a minor comment.

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/provider/PinotConfigProvider.java (hunk 1 above)
Comment (npawar) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/provider/PinotConfigProvider.java:28]: makes sense, yeah i don't see us extending this to anything more (at the most to segment metadata). Added to javadocs

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/provider/PinotConfigProvider.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  /**
+   * Gets the tableConfig
+   */
+  TableConfig getTableConfig(String tableNameWithType);
+
+  /**
+   * Registers the {@link TableConfigChangeListener} and notifies it whenever any changes (addition, update, removal)
+   * to any of the table configs are detected
+   */
+  void registerTableConfigChangeListener(TableConfigChangeListener tableConfigChangeListener);
```
Comment (mcvsubbu) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/provider/PinotConfigProvider.java:39]: Let us say a user of this interface wants to track the changes in table config. Looks like they need to get table config first and then register to listen to changes.

Is there a race condition here? (changes may happen before they register, for instance).

Does'nt it make sense to return the table config at the time of registration?

Same with schema I suppose

Comment (npawar) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/provider/PinotConfigProvider.java:39]: They can directly register the listener. No need to get table config first.

Comment (npawar) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/provider/PinotConfigProvider.java]: On second thought, that does make it better for the specific case you're talking about. changed it.

File: pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java [hunk 3]
```diff
@@ -142,24 +153,38 @@ public String getActualTableName(String caseInsensitiveTableName) {
...
   @Nullable
+  @Override
   public Schema getSchema(String rawTableName) {
     String schemaName = _schemaNameMap.getOrDefault(rawTableName, rawTableName);
     SchemaInfo schemaInfo = _schemaInfoMap.get(schemaName);
     return schemaInfo != null ? schemaInfo._schema : null;
   }
 
+  @Override
+  public List<Schema> registerSchemaChangeListener(SchemaChangeListener schemaChangeListener) {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java:179]: should be synchronized so that we don't collide with notification that may come in another thread.

File: pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java [hunk 4]
```diff
@@ -142,24 +153,38 @@ public String getActualTableName(String caseInsensitiveTableName) {
    * Returns the table config for the given table, or {@code null} if it does not exist.
    */
   @Nullable
+  @Override
   public TableConfig getTableConfig(String tableNameWithType) {
     return _tableConfigMap.get(tableNameWithType);
   }
 
+  @Override
+  public List<TableConfig> registerTableConfigChangeListener(TableConfigChangeListener tableConfigChangeListener) {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java:162]: should be synchronized so that we don't collide with the notification thread

File: pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java [hunk 5]
```diff
@@ -45,10 +52,11 @@
 
 
 /**
+ * An implementation of {@link PinotConfigProvider}
  * The {@code TableCache} caches all the table configs and schemas within the cluster, and listens on ZK changes to keep
  * them in sync. It also maintains the table name map and the column name map for case-insensitive queries.
  */
-public class TableCache {
+public class TableCache implements PinotConfigProvider {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java:59]: Maybe good to rename this as DefaultConfigProvider (or some such), but perhaps that is better done in another PR.

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/provider/PinotConfigProvider.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+ */
+package org.apache.pinot.spi.config.provider;
+
+import java.util.List;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.Schema;
+
+
+/**
+ * An interface for the provider of pinot table configs and schema
```
Comment (jackjlli) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/provider/PinotConfigProvider.java:27]: It seems we've already had a `TableConfigs` class that covers table schema and config. Should we consider merging these two?

File: pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java (hunk 4 above)
Comment (jackjlli) [inline pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java:162]: I guess there will be another PR that includes the code to explicitly invoke these two methods?

File: pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java [hunk 7]
```diff
@@ -57,10 +65,13 @@
   private static final String LOWER_CASE_OFFLINE_TABLE_SUFFIX = "_offline";
   private static final String LOWER_CASE_REALTIME_TABLE_SUFFIX = "_realtime";
 
+  private final Set<TableConfigChangeListener> _tableConfigChangeListeners = new HashSet<>();
```
Comment (jackjlli) [inline pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java:68]: It'd be good to use `ConcurrentHashMap.newKeySet()` here to align with `_tableConfigMap`.

File: pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java [hunk 8]
```diff
@@ -57,10 +65,13 @@
...
+  private final Set<SchemaChangeListener> _schemaChangeListeners = new HashSet<>();
```
Comment (jackjlli) [inline pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java:69]: Same here. It'd be good to use `ConcurrentHashMap.newKeySet()`.

File: pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java (hunk 4 above)
Comment (snleee) [reply pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java:162]: +1

**Review by snleee — APPROVED** [review]
Please double-check on the synchronization between `registerTableConfigChangeListener` and the notification thread as @mcvsubbu  mentioned. Otherwise, LGTM.

File: pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java (hunk 4 above)
Comment (npawar) [reply pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java:162]: i did not follow how will adding `synchronized` here will help? or do you mean add it here and also to `notifyTableConfigChangeListeners` method?

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/provider/PinotConfigProvider.java (hunk 6 above)
Comment (npawar) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/provider/PinotConfigProvider.java:27]: that's a good point. but i think that can be a new method on this interrface when someone needs to use that. basic table config and schema seem like a good starting point

File: pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java (hunk 4 above)
Comment (npawar) [reply pinot-common/src/main/java/org/apache/pinot/common/config/provider/TableCache.java:161]: added synchronized
